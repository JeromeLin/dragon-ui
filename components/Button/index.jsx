import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Icon from '../Icon';

class Button extends Component {
  render() {
    const { props } = this;
    const {
      type,
      theme,
      size,
      isBlock,
      isRadius,
      isRound,
      isCircle,
      isActive,
      isFocus,
      isDisabled,
      isLoading,
      className,
      onClick,
      children,
      style
    } = props;
    const disabled = 'disabled' in props || isDisabled;

    const classes = classnames({
      'ui-button': true,
      block: 'block' in props || isBlock,
      radius: 'radius' in props || isRadius,
      round: 'round' in props || isRound,
      circle: 'circle' in props || isCircle,
      active: 'active' in props || isActive,
      focus: 'focus' in props || isFocus,
      disabled,
      [`theme-${theme}`]: !!theme,
      [`size-${size}`]: !!size,
      [className]: !!className
    });

    let textContent =
      'loading' in props || isLoading ? (
        <span>
          <Icon type="loading" className="rotate360" /> {children}
        </span>
      ) : (
        children
      );
    return (
      <button
        type={type}
        className={classes}
        style={style}
        disabled={disabled}
        onClick={e => !disabled && onClick(e)}
      >
        {textContent}
      </button>
    );
  }
}

Button.propTypes = {
  type: PropTypes.string,
  theme: PropTypes.oneOf(['default', 'info', 'success', 'warning', 'error']),
  size: PropTypes.oneOf(['xl', 'lg', 'sm', 'xs']),
  isBlock: PropTypes.bool,
  isRadius: PropTypes.bool,
  isRound: PropTypes.bool,
  isCircle: PropTypes.bool,
  isActive: PropTypes.bool,
  isFocus: PropTypes.bool,
  isDisabled: PropTypes.bool,
  isLoading: PropTypes.bool,
  className: PropTypes.string,
  style: PropTypes.objectOf(PropTypes.any),
  onClick: PropTypes.func,
};

Button.defaultProps = {
  type: 'button',
  theme: 'default',
  size: null,
  isBlock: false,
  isRadius: false,
  isRound: false,
  isCircle: false,
  isActive: false,
  isFocus: false,
  isDisabled: false,
  isLoading: false,
  className: null,
  style: {},
  onClick: () => {}
};

export default Button;
