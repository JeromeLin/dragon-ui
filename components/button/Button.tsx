import React, { Component, ButtonHTMLAttributes, AnchorHTMLAttributes, ReactNode } from 'react';
import classnames from 'classnames';
import ButtonProps, { ButtonType } from './PropsType';
import Icon from '../icon';
import ButtonGroup from './ButtonGroup';

interface ButtonPropsIF extends ButtonProps {
  htmlType: ButtonType;
}


interface ButtonIF {
  anchor: AnchorHTMLAttributes<HTMLAnchorElement> & ButtonProps;
  button: ButtonHTMLAttributes<HTMLButtonElement> & ButtonPropsIF;
}

function isAnchorProps(props: any):
  props is AnchorHTMLAttributes<HTMLAnchorElement> & ButtonProps {
  return {}.hasOwnProperty.call(props, 'href');
}

// function isButtonProps(props: Button['props']): props is (Readonly<ButtonIF['button']> & Readonly<{ children?: ReactNode; }>) {
//   return !{}.hasOwnProperty.call(props, 'href');
// }


class Button<T extends 'button' | 'anchor' = 'button'> extends Component<ButtonIF[T]> {
  static Group = ButtonGroup;

  static defaultProps = {
    prefixCls: 'zw-button',
    theme: 'default',
    shape: 'radius',
    ghost: false,
    size: null,
    loading: false,
    disabled: false,
    block: false,
    htmlType: 'button',
  };

  onClick = (e: any) => {
    const { disabled, loading, onClick } = this.props;
    if (disabled === true || loading === true) {
      return;
    }
    if (onClick) {
      onClick(e);
    }
  };

  renderAnchor(props: Readonly<ButtonIF['anchor'] & { children?: ReactNode }>, classes: string, textContent: ReactNode) {
    const {
      prefixCls,
      htmlType,
      type,
      size,
      block,
      shape,
      disabled,
      ghost,
      loading,
      className,
      onClick,
      children,
      theme,
      active,
      ...restAnchorProps
    } = props;
    return (
      <a
        {...restAnchorProps}
        className={classes}
        onClick={this.onClick}
      >
        {textContent}
      </a>
    );
  }

  renderButton(props: Readonly<ButtonIF['button']> & Readonly<{ children?: ReactNode }>, classes: string, textContent: ReactNode) {
    const {
      prefixCls,
      htmlType,
      type,
      size,
      block,
      shape,
      ghost,
      loading,
      className,
      onClick,
      children,
      theme,
      active,
      ...restButtonProps
    } = props;
    return (
      <button
        {...restButtonProps}
        type={htmlType}
        className={classes}
        onClick={this.onClick}
      >
        {textContent}
      </button>
    );
  }

  render() {
    const {
      prefixCls,
      size,
      block,
      shape,
      disabled,
      ghost,
      loading,
      className,
      theme,
      active,
      focus,
      children,
    } = this.props;

    const classes = classnames(prefixCls, className, {
      [`${prefixCls}--${theme}`]: theme,
      [`${prefixCls}--${size}`]: size,
      [`${prefixCls}--block`]: block,
      [`${prefixCls}--${shape}`]: shape,
      [`${prefixCls}--active`]: active,
      [`${prefixCls}--focus`]: focus,
      [`${prefixCls}--disabled`]: disabled,
      [`${prefixCls}--loading`]: loading,
      [`${prefixCls}--ghost`]: ghost,
    });

    const textContent = loading
      ? (
        <React.Fragment>
          <Icon type="loading" className="rotate360" />
          &nbsp;&nbsp;
          {children}
        </React.Fragment>
      )
      : children;

    const allProps = this.props as any;
    if (isAnchorProps(allProps)) {
      return this.renderAnchor(allProps, classes, textContent);
    }
    return this.renderButton(allProps, classes, textContent);
  }
}

export default Button;
