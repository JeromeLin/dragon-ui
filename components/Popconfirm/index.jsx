import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Popover from '../Popover';
import Button from '../Button';

const noop = () => {};

class Popconfirm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      visible: false
    };
  }

  componentWillReceiveProps(nextProps) {
    if (this.state.visible !== nextProps.visible) {
      this.setState({
        visible: !!nextProps.visible
      });
    }
  }

  handleCancel() {
    const { onCancel } = this.props;
    this.setState({
      visible: false
    });
    onCancel();
  }

  handleConfirm() {
    const { onOk } = this.props;
    this.setState({
      visible: false
    });
    onOk();
  }

  render() {
    const { visible } = this.state;
    const {
      children,
      content,
      prefixCls,
      className,
      trigger,
      direction,
      okText,
      cancelText
    } = this.props;

    // eslint-disable-next-line
    const popContent = () => {
      return (
        <div>
          {content}
          <div className={`${prefixCls}-actions`}>
            <Button onClick={() => this.handleCancel()}>{cancelText}</Button>
            <Button theme="info" onClick={() => this.handleConfirm()}>
              {okText}
            </Button>
          </div>
        </div>
      );
    };

    return (
      <Popover
        content={popContent}
        visible={visible}
        direction={direction}
        prefixCls={prefixCls}
        className={className}
        trigger={trigger}
      >
        {children}
      </Popover>
    );
  }
}

Popconfirm.propTypes = {
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  visible: PropTypes.bool,
  trigger: PropTypes.oneOf(['click', 'hover']),
  direction: PropTypes.oneOf([
    'topLeft',
    'top',
    'topRight',
    'rightTop',
    'right',
    'rightBottom',
    'bottomLeft',
    'bottom',
    'bottomRight',
    'leftTop',
    'left',
    'leftBottom'
  ]),
  onCancel: PropTypes.func,
  onOk: PropTypes.func,
  okText: PropTypes.string,
  cancelText: PropTypes.string,
  content: PropTypes.node
};

Popconfirm.defaultProps = {
  prefixCls: 'ui-popover',
  className: null,
  visible: false,
  trigger: 'click',
  direction: 'bottomLeft',
  onCancel: noop,
  onOk: noop,
  okText: '确认',
  cancelText: '取消',
  content: null
};

export default Popconfirm;
