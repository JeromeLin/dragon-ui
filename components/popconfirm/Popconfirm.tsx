import React, { Component } from 'react';
import classnames from 'classnames';
import Popover from '../popover';
import Button from '../button';
import PopconfirmProps from './PropsType';
import Icon from '../icon';

interface PopconfirmStates {
  visible: boolean;
}

class Popconfirm extends Component<PopconfirmProps, PopconfirmStates> {
  static defaultProps: PopconfirmProps = {
    prefixCls: 'zw-popconfirm',
    trigger: 'click',
    direction: 'top',
    icon: <Icon type="warning-round-fill" theme="warning" />,
  };

  state: PopconfirmStates = {
    visible: false,
  };

  static getDerivedStateFromProps(props: PopconfirmProps, state: PopconfirmStates) {
    if ('visible' in props && props.trigger === 'manual') {
      return {
        ...state,
        visible: props.visible,
      };
    }
    return null;
  }

  private isManual = (): boolean => {
    const { trigger } = this.props;
    return trigger === 'manual';
  };

  private onVisibleChange = (visible: boolean): void | boolean => {
    if (visible === this.state.visible) return false;
    this.setState({ visible });
  };

  private handleClose = (): void => {
    !this.isManual() && this.setState({ visible: false });
  };

  private handleCancel = (): void => {
    const { onCancel } = this.props;
    typeof onCancel === 'function' && onCancel();
    this.handleClose();
  };

  private handleConfirm = (): void => {
    const { onOk } = this.props;
    typeof onOk === 'function' && onOk();
    this.handleClose();
  };

  render() {
    const {
      prefixCls,
      className,
      children,
      content,
      cancelText,
      okText,
      locale,
      icon,
      ...others
    } = this.props;
    const { visible } = this.state;
    const cls = classnames(prefixCls, className);

    console.log(cancelText, locale);
    const popContent: React.ReactNode = (
      <>
        <div className={`${prefixCls}__message`}>
          {icon}
          <div className={`${prefixCls}__title`}>{content}</div>
        </div>
        <div className={`${prefixCls}__buttons`}>
          <Button size="sm" onClick={this.handleCancel}>
            {cancelText || locale!.cancelText}
          </Button>
          <Button size="sm" theme="primary" onClick={this.handleConfirm}>
            {okText || locale!.okText}
          </Button>
        </div>
      </>
    );

    return (
      <Popover
        className={cls}
        content={popContent}
        {...others}
        visible={visible}
        onVisibleChange={this.onVisibleChange}
      >
        {children}
      </Popover>
    );
  }
}

export default Popconfirm;
