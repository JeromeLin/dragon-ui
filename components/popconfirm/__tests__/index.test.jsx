import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Popconfirm from '../index';
import Button from '../../button';

if (global.document) {
  document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
      nodeName: 'BODY',
      ownerDocument: document,
    },
  });
}

describe('PopConfirm', () => {
  it('renders PopConfirm correctly', () => {
    const wrapper = render(
      <div>
        <Popconfirm content="确认删除吗" direction="top">
          <Button>Delete</Button>
        </Popconfirm>
      </div>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('check `onVisibleChange` prop', () => {
    const onVisibleChange = jest.fn();

    const wrapper = mount(
      <Popconfirm content="hello" trigger="click" onVisibleChange={onVisibleChange}>
        <div id="hello">Hello world!</div>
      </Popconfirm>,
    );
    const div = wrapper.find('#hello');
    div.simulate('click');
    setTimeout(() => {
      expect(onVisibleChange).toHaveBeenLastCalledWith(true);
    });

    div.simulate('click');
    setTimeout(() => {
      expect(onVisibleChange).toHaveBeenLastCalledWith(false);
    });
  });

  it('check onOk function is call', () => {
    const ok = jest.fn();
    const wrapper = mount(
      <Popconfirm content="hello" trigger="click" onOk={ok}>
        <div id="hello">Hello world!</div>
      </Popconfirm>,
    );
    const hello = wrapper.find('#hello');
    hello.simulate('click');
    const div = wrapper.find('.zw-button--primary');
    div.simulate('click');

    expect(ok.mock.calls.length).toBe(1);
  });

  it('check onCancel function is call', () => {
    const onCancel = jest.fn();
    const wrapper = mount(
      <Popconfirm content="hello" trigger="click" onCancel={onCancel}>
        <div id="hello">Hello world!</div>
      </Popconfirm>,
    );
    const hello = wrapper.find('#hello');
    hello.simulate('click');
    const div = wrapper.find('.zw-button--default');
    div.simulate('click');

    expect(onCancel.mock.calls.length).toBe(1);
  });
});
