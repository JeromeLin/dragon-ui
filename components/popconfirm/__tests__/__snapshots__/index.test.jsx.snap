// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopConfirm renders PopConfirm correctly 1`] = `<div />`;
